docs/zephyr_02_functional_requirements.sdoc

```diff
@@ -1317,6 +1317,248 @@ As a Zephyr OS user I want to... ???,
+UID: ZEP-129
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore dynamic initialization
+STATEMENT: >>>
+Zephyr Counting Semaphores shall expose an interface to dynamically initialize a Counting Semaphore which requires the following information:
+
+- initial count of the Counting Semaphore; and
+- limit of the Counting Semaphore.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-130
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphores Static Object Init Failures Cause Build Errors
+STATEMENT: >>>
+While a Counting Semaphore is statically allocated,
+
+If the Counting Semaphore is configured with an invalid set of initialization parameter values,
+
+the Zephyr Build System shall fail the build and report the point of failure.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-131
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Dynamic Object Init Failures Reports Failure Reason
+STATEMENT: >>>
+While a Counting Semaphore is dynamically allocated,
+
+If the Counting Semaphore initialization is configured with an invalid set of initialization parameter values,
+
+the Counting Semaphore initialization shall report the reason for the failure.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-132
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Exposes Current Count
+STATEMENT: >>>
+When the Counting Semaphore current count is requested,
+
+Zephyr shall report the current count of the Counting Semaphore.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-133
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Current Count Request Invalid Parameter is Catastrophic Error
+STATEMENT: >>>
+When the Counting Semaphore current count is requested,
+
+If the Counting Semaphore current count parameters are invalid,
+
+Zephyr shall report a catastrophic failure.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-134
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphores Exposes Ability to Reset Count to Zero
+STATEMENT: >>>
+When the Counting Semaphore reset is requested,
+
+Zephyr shall do all of the following:
+
+- set to zero the current count of the Counting Semaphore, and
+- abort all other outstanding attempts to interact with the Counting Semaphore.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-135
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Reset Request Invalid Parameter is Catastrophic Error
+STATEMENT: >>>
+When the Counting Semaphore reset is requested,
+
+If the Counting Semaphore reset parameters are invalid,
+
+Zephyr shall report a catastrophic failure.
+<<<
+
+[REQUIREMENT]
+UID: ZEP-136
+STATUS: Draft
+TYPE: Functional
+COMPONENT: Counting Semaphores
+TITLE: Counting Semaphore Give unblocks pending Take by another thread
+STATEMENT: >>>
+While a thread is blocked waiting to Take from this Counting Semaphore,
+
+When the Counting Semaphore Give is requested,
```

Comment:
**suggestion:** It would be easier to read, "When a running task gives the semaphore ..."

Also the statement as such is most likely inprecise. An arbitrary number of tasks can wait on a semaphore. When the semaphore is given, only the highest priority (that's where determinism kicks in) thread is unblocked.

Reply:
Agreed.  Thank you for the suggestion.